File: src/core/safety/permissionGrid.ts
/**
 * chorus H10 (resolved AMBIGUO) — explicit per-(scope, resource)
 * permission grid.
 *
 * Three scopes ("once" / "session" / "always") cross N resource
 * patterns. The UI renders it as a table; the underlying storage is
 * flat so export and replay are trivial. The same resource pattern in
 * different scopes is independent — `cmd.exec.git` allowed once but
 * denied for the session is a coherent and supported configuration.
 *
 * Setting state="unset" is equivalent to clearing the cell so callers
 * do not have to know whether a cell was previously set. `get` always
 * returns a non-null `PermissionState` ("unset" when no row exists).
 */

export type PermissionScope = "once" | "session" | "always";
export type PermissionState = "allow" | "deny" | "unset";

export interface PermissionRow {
  scope: PermissionScope;
  resource: string;
  state: PermissionState;
}

export class PermissionGrid {
  private rows = new Map<string, PermissionState>();

  private key(scope: PermissionScope, resource: string): string {
    return `${scope}::${resource}`;
  }

  set(
    scope: PermissionScope,
    resource: string,
    state: PermissionState,
  ): void {
    if (state === "unset") {
      this.rows.delete(this.key(scope, resource));
    } else {
      this.rows.set(this.key(scope, resource), state);
    }
  }

  get(scope: PermissionScope, resource: string): PermissionState {
    return this.rows.get(this.key(scope, resource)) ?? "unset";
  }

  exportRows(): PermissionRow[] {
    const out: PermissionRow[] = [];
    for (const [k, state] of this.rows) {
      // Resource may contain "::" so we split on the FIRST occurrence only.
      const sep = k.indexOf("::");
      const scope = k.slice(0, sep) as PermissionScope;
      const resource = k.slice(sep + 2);
      out.push({ scope, resource, state });
    }
    return out;
  }
}

File: crates/apohara-safety/src/permission_grid.rs
//! Explicit per-(scope, resource) permission grid — ports
//! `src/core/safety/permissionGrid.ts` (chorus H10).

use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, Hash)]
#[serde(rename_all = "lowercase")]
pub enum PermissionScope {
    Once,
    Session,
    Always,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum PermissionState {
    Allow,
    Deny,
    Unset,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct PermissionRow {
    pub scope: PermissionScope,
    pub resource: String,
    pub state: PermissionState,
}

#[derive(Debug, Default)]
pub struct PermissionGrid {
    rows: HashMap<(PermissionScope, String), PermissionState>,
}

impl PermissionGrid {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn set(&mut self, scope: PermissionScope, resource: &str, state: PermissionState) {
        if matches!(state, PermissionState::Unset) {
            self.rows.remove(&(scope, resource.to_string()));
        } else {
            self.rows.insert((scope, resource.to_string()), state);
        }
    }

    pub fn get(&self, scope: PermissionScope, resource: &str) -> PermissionState {
        self.rows
            .get(&(scope, resource.to_string()))
            .copied()
            .unwrap_or(PermissionState::Unset)
    }

    pub fn export_rows(&self) -> Vec<PermissionRow> {
        self.rows
            .iter()
            .map(|((scope, resource), state)| PermissionRow {
                scope: *scope,
                resource: resource.clone(),
                state: *state,
            })
            .collect()
    }
}
